### DISCUSSION TITLE
Vocês acham que é possível utilizar o pynubank para todo dia registrar novas transações em uma base de dados?

### DISCUSSION THREAD

**Discussion by davidborghi** [andreroggeri/pynubank#395] [Q&A] [answered]
Olá, espero que estejam bem! 

Parabéns e obrigado pelo projeto! 

Apenas uma breve observação, é inacreditável como os bancos ainda não implementaram API's completas, funcionais e bem documentadas, assim como acho bem frustrante que não possamos baixar extratos em .json que facilitaria muito nossa organização financeira. 

Atualmente uso **[SeaTable](https://seatable.io/),** é uma alternativa a Airtable, a minha intenção é alimentar uma base com meus gastos e receitas da nuconta, porém, para deixá-los organizados tenho feito da seguinte forma: 

1. Baixar o extrato mensal.
2. Separar os campos adequadamente no arquivo .csv.
3. Aplicar expressão regular.
4. Upload para Seatable.

Vocês acham que seria possível usar o projeto para todos os dias ele rodar e atualizar uma base da dados ou um arquivo .csv que posteriormente faria a atualização via n8n (provavelmente) da base de dados no Seatable?

Caso não, vocês teriam alguma outra sugestão de fluxo mais fácil ou adequado?

Obrigado pela atenção e disposição!

Desculpe-me pelo incômodo!

Tudo de bom à todos!

**Comment by andreroggeri** [comment, marked as answer]
Pois é, eu achei que o open finance ajudaria nesse sentido mas não 😢 .

Sobre a sua idéia, funciona sim, inclusive eu fiz essa lib justamente pra isso. 
Eu [sincronizo](https://github.com/andreroggeri/infrastructure/blob/main/nas/compose/cron.docker-compose.yml#L51) todas as transações duas vezes por dia no [YNAB](https://www.ynab.com/) que é a ferramenta que uso pra gestão financeira.

**Comment by davidborghi** [comment]
Obrigado pela tão rápida resposta! 

Que boa notícia! 

Sei que há inúmeras formas de se fazer isso, queria fazer da forma mais simples possível para evitar manutenção, é que tô com muita coisa pra fazer na vida e não queria ter mais uma pra ter que "monitorar" com frequência, mas ao mesmo tempo não ter a vida financeira organizada da forma como gostaria tem me trazido preocupação.

Eu tô gastando muito tempo usando parse pra deixar os dados como gostaria e ainda não tá ficando da forma ideal.

Pensei em usar o n8n, o que você acha?

Caso prefira outra abordagem, por favor, fique à vontade para dizer.

Pelo que vi você está usando docker, tenho bem pouca experiência com ele, mesmo sendo algo relativamente "básico" no mundo da programação.

O meu intuito seria capturar diariamente as transações, formatá-las e levar para o SeaTable, optei por usar ele ao invés do YNAB pois eu uso-o para guardar todas as outras coisas importantes, por mais que eu perca um pouco de personalização, mantenho tudo em um lugar só, mas é uma alternativa bastante interessante! :)

1. Com relação à todas as requisições possíveis, você tem uma lista ou as que estão nos "examples" já são todas?
2. Há como capturar as transações com os mesmos campos que aparecem nos comprovantes?
3. É possível capturar todas as transações em ordem cronológica mesmo sendo de tipos diferentes (crédito, débito, transferência, etc.)?[
4. Pra você, o que faz mais sentido, capturar todas as transações excluindo o cartão de crédito e apenas usando o "pagamento de fatura" como referência ou você acha que vale a pena gastar mais tempo e ter as informações das compras no crédito detalhadas também?

Desculpe-me pela quantidade de perguntas, fiquei bastante animado com a possibilidade de dar certo! 

Não preocupe-se caso não possa responder!

Muito obrigado pela atenção e disposição!

Desculpe-me pelo incômodo!

**Comment by andreroggeri** [comment]
> Pensei em usar o n8n, o que você acha?

Não conhecia, mas pelo que eu bati o olho na documentação, parece atender as suas necessidades.

> Com relação à todas as requisições possíveis, você tem uma lista ou as que estão nos "examples" já são todas?

Não estão todas no examples, mas acredito que o `get_account_statements_paginated` e o `get_card_statements` devem atender 95% da sua necessidade

> Há como capturar as transações com os mesmos campos que aparecem nos comprovantes?

Sim, mas nem todas as informações são retornadas nos métodos que trazem a lista, então dependendo do que você for precisar, será necessário fazer mais uma chamada por transação (O que pode ser complicado pois o Nubank tem rate-limiting não documentado)

> É possível capturar todas as transações em ordem cronológica mesmo sendo de tipos diferentes (crédito, débito, transferência, etc.)?

Todas as transações tem data/hora, se for juntar as duas listas (transações da Nuconta + cartão de crédito) pode ser necessário fazer a ordenação após juntar

> Pra você, o que faz mais sentido, capturar todas as transações excluindo o cartão de crédito e apenas usando o "pagamento de fatura" como referência ou você acha que vale a pena gastar mais tempo e ter as informações das compras no crédito detalhadas também?

Depende de como você controla os seus gastos. No meu caso, a maioria dos gastos estão no cartão de crédito, se eu não importar cada transação separadamente, eu não vou conseguir categorizar / orçar corretamente.

**Reply by davidborghi** [reply]
Obrigado mais uma vez pela tão rápida resposta!

Desculpe-me pelo excesso de perguntas, caso não possa responder por qualquer motivo, por favor, fique à vontade e entenderei sem nenhum problema.

&nbsp;

****
> Não conhecia, mas pelo que eu bati o olho na documentação, parece atender as suas necessidades.

Opto por usar ele de vez em quando pois ele é self-hosted e dá pra segmentar os fluxos, é bastante interessante, mas você acha que teria alguma abordagem mais eficiente? 

Se sim, por favor, fique à vontade para dizer.

> Não estão todas no examples, mas acredito que o get_account_statements_paginated e o get_card_statements devem atender 95% da sua necessidade

Entendido, obrigado!

> Sim, mas nem todas as informações são retornadas nos métodos que trazem a lista, então dependendo do que você for precisar, será necessário fazer mais uma chamada por transação (O que pode ser complicado pois o Nubank tem rate-limiting não documentado)

Entendido, obrigado pelo alerta, na busca pelas informações (antes de abrir o tópico) acabei lendo algumas discussões aqui no projeto sobre isso. :)

> Todas as transações tem data/hora, se for juntar as duas listas (transações da Nuconta + cartão de crédito) pode ser necessário fazer a ordenação após juntar

Imagino que talvez fosse mais fácil apenas organizar direto lá no Seatable, pois eles tem filtros, grupos, ordenação etc. o que você acha?

Eu penso que, por princípio, o correto seria tratar antes de sincronizar, mas, como falei, tenho tentado balancear o possível com a maior probabilidade de ser executável e com menor manutenção, mas adoraria saber sua opinião sincera sobre isso.

> Depende de como você controla os seus gastos. No meu caso, a maioria dos gastos estão no cartão de crédito, se eu não importar cada transação separadamente, eu não vou conseguir categorizar / orçar corretamente.

Eu acabo utilizando mais o débito e pix, tanto para deixar o crédito como emergência e para evitar colocar coisas que são de recorrência mensal (água, luz, mercado etc.). 

****
&nbsp;

O meu sonho era conseguir capturar as transações de forma completa, com todas informações que aparecem no comprovante de transferência, tal como descrevi **[aqui](https://comunidade.nubank.com.br/t/queremos-ouvir-sua-opini%C3%A3o-sobre-a-conta-do-nubank/407415/395)** e **[aqui](https://comunidade.nubank.com.br/t/queremos-ouvir-sua-opini%C3%A3o-sobre-a-conta-do-nubank/407415/397).**

1. Para eu evitar lhe fazer tantas perguntas, qual a melhor forma de eu tentar me guiar para entender as possibilidades do que posso fazer com o projeto?

2. Se eu quisesse capturar tanto as transações de crédito como nuconta, há como fazer isso em um script só ou obrigatoriamente eu precisaria fazer dois fluxos?

3. Já pensei em qual lógica adotar para o fluxo, mas como não tenho muita experiência, fico sempre achando que estou escolhendo a abordagem errada, qual é a lógica mais adequada nesse caso?

    3.1 Apenas usar alguma função ("today") para capturar as transações do dia?

4. Caso não seja possível capturar as informações de maneira completa (como nos casos que descrevi nos links), é possível capturar as informações apenas com os campos que desejar, ou seja, "fazer o tratamento" já na requisição?

5. Pelo que conheço do YNAB, o princípio é ter um orçamento programado para todas as coisas, atualmente, no seu fluxo, você importa os recebimentos diariamente e ele já segmenta automaticamente ou apenas importa os recebimentos e você faz manualmente? O mesmo se aplica aos gastos também?

Obrigado pela atenção e disposição!

Desculpe-me pelo incômodo!

**Reply by andreroggeri** [reply]
É possível fazer o que você quer, mas talvez antes de chegar no "perfeito" seria legal ir fazendo as coisas de maneira incremental. Assim você pode validar o que funciona e o que não.

> Para eu evitar lhe fazer tantas perguntas, qual a melhor forma de eu tentar me guiar para entender as possibilidades do que posso fazer com o projeto?

Você pode olhar as informações que são retornadas nos métodos `get_account_statements_paginated` e `get_card_statements`. Assim você consegue ter uma idéia das informações que estão disponíveis.

> Se eu quisesse capturar tanto as transações de crédito como nuconta, há como fazer isso em um script só ou obrigatoriamente eu precisaria fazer dois fluxos?

É simples como chamar os dois métodos no mesmo script, por ex:

```py
from pynubank import Nubank

nu = Nubank()
account_statements = nu.get_account_statements_paginated()
card_statements = nu.get_card_statements()

# Implementar funções abaixo
import_account_statements(account_statements)
import_card_statements(card_statements)
```

> Caso não seja possível capturar as informações de maneira completa (como nos casos que descrevi nos links), é possível capturar as informações apenas com os campos que desejar, ou seja, "fazer o tratamento" já na requisição?

Em teoria, tudo que retorna no app pode ser recuperado aqui (talvez não tenha um método implementado ainda)


> Pelo que conheço do YNAB, o princípio é ter um orçamento programado para todas as coisas, atualmente, no seu fluxo, você importa os recebimentos diariamente e ele já segmenta automaticamente ou apenas importa os recebimentos e você faz manualmente? O mesmo se aplica aos gastos também?

A categorização eu faço manualmente, sempre que o meu script roda, eu recebo uma notificação no celular para me lembrar de fazer a categorização. Exceto no caso de transações de um estabelecimento que já existe no YNAB, nesse caso ele usa a mesma categoria da transação anterior. O mesmo vale pra gastos e recebimentos (A única diferença pro YNAB é se o valor é positivo ou negativo hehe)
![image](https://user-images.githubusercontent.com/6232673/235354936-bb86a14d-26c6-4a9a-b821-7bca6f88a594.png)

**Reply by davidborghi** [reply]
> É possível fazer o que você quer, mas talvez antes de chegar no "perfeito" seria legal ir fazendo as coisas de maneira incremental. Assim você pode validar o que funciona e o que não.

Obrigado pela resposta e pela dica, acabei fazendo dessa forma, transformei-os em um dataframe e salvei em .csv. 

Usei o `account_statements = nu.get_account_statements()` para NuConta e o `card_statements = nu.get_card_statements()` para o cartão de crédito, o problema é que os arquivos gerados não possuem os mesmos campos nem seguem a mesma ordem, mas imagino que seja possível resolver já passando os campos que deseja-se extrair e a ordem, correto?

> É simples como chamar os dois métodos no mesmo script, por ex:
> 
> ```python
> from pynubank import Nubank
> 
> nu = Nubank()
> account_statements = nu.get_account_statements_paginated()
> card_statements = nu.get_card_statements()
> 
> # Implementar funções abaixo
> import_account_statements(account_statements)
> import_card_statements(card_statements)
> ```

Tentei fazer o mesmo para o feed paginado, usei o seu código de exemplo, mas os arquivos .json gerados foram apenas exemplificativos e não relacionados as minhas transações de fato, nessa opção de feed paginado, teoricamente é para vir toda transação (pix, débito, crédito, reembolso etc.)?

Pergunto isso, pois, ao gerar os arquivos .csv do cartão e da nuconta, as transações usando cartão de débito não apareceram.

Mudei para gerar apenas a primeira página e novamente apenas foi apresentado os dados exemplificativos, usei o seguinte código: 

```python
import json
from pynubank import Nubank, MockHttpClient

nu = Nubank(MockHttpClient())
nu.authenticate_with_cert('cpf', 'pass', 'cert')

# Obter apenas a primeira página de transações
feed = nu.get_account_feed()

# Salvar a primeira página em um arquivo JSON
with open('feed_1.json', 'w') as f:
    json.dump(feed, f)
```

> A categorização eu faço manualmente, sempre que o meu script roda, eu recebo uma notificação no celular para me lembrar de fazer a categorização. Exceto no caso de transações de um estabelecimento que já existe no YNAB, nesse caso ele usa a mesma categoria da transação anterior. O mesmo vale pra gastos e recebimentos (A única diferença pro YNAB é se o valor é positivo ou negativo hehe)

Obrigado pela explicação! 

Atualmente você considera que o YNAB lhe ajuda à ficar mais organizado mesmo o controle sendo feito de forma assíncrona, ou seja, a separação dos valores ocorrer apenas no YNAB e você não ver necessariamente isso direto na sua conta?

&nbsp;

Muito obrigado por toda atenção e disposição!

Desculpe-me pelo incômodo!

Tenha um excelente final de semana!

**Reply by andreroggeri** [reply]
> Pergunto isso, pois, ao gerar os arquivos .csv do cartão e da nuconta, as transações usando cartão de débito não apareceram.

Mudei para gerar apenas a primeira página e novamente apenas foi apresentado os dados exemplificativos, usei o seguinte código:

É porque você tá usando o `MockHttpClient` no construtor. Quando você faz isso, só vai receber exemplos pq não está chamando a API do Nubank.

Basta mudar para ficar assim:

```
import json
from pynubank import Nubank, MockHttpClient

# nu = Nubank(MockHttpClient())
nu = Nubank() # Sem o MockHttpCLient :)
nu.authenticate_with_cert('cpf', 'pass', 'cert')

# Obter apenas a primeira página de transações
feed = nu.get_account_feed()

# Salvar a primeira página em um arquivo JSON
with open('feed_1.json', 'w') as f:
    json.dump(feed, f)
```

**Reply by andreroggeri** [reply]
> Atualmente você considera que o YNAB lhe ajuda à ficar mais organizado mesmo o controle sendo feito de forma assíncrona, ou seja, a separação dos valores ocorrer apenas no YNAB e você não ver necessariamente isso direto na sua conta?

O controle não é feito de maneira assincrona, o YNAB representa exatamente o estado atual de todas as contas num lugar só, mas além disso esse dinheiro está categorizado. Por exemplo, tenho R$ 500 reais na conta e estou com vontade de pedir um ifood. Se eu olho o saldo da conta, eu tenho dinheiro suficiente para isso, mas eu tenho que lembrar de todas as contas que eu ainda preciso pagar. Então se eu olhar pro meu orçamento no YNAB eu consigo ter uma visão clara do que eu posso / não posso fazer.

![image](https://user-images.githubusercontent.com/6232673/236510826-e53dd06b-cb2a-4fd7-8ca9-7beacb0cfd2b.png)

Nesse exemplo ☝️  eu tinha alocado 100 reais pra comer fora e já gastei 99, então não posso mais gastar nessa categoria.

![image](https://user-images.githubusercontent.com/6232673/236511014-e0e1ab39-f24d-437b-a5f9-a75031eee03c.png)

Olhando para os gastos obrigatórios, eu tenho 500 reais alocados que são referentes ao aluguel.


Então o saldo da conta acaba sendo irrelevante, pq conforme eu vou recebendo dinheiro, ele vai sendo alocado e o orçamento é a minha fonte da verdade pra saber quanto eu posso gastar com cada coisa.

**Reply by davidborghi** [reply]
> Mudei para gerar apenas a primeira página e novamente apenas foi apresentado os dados exemplificativos, usei o seguinte código:
> 
> É porque você tá usando o `MockHttpClient` no construtor. Quando você faz isso, só vai receber exemplos pq não está chamando a API do Nubank.
> 
> Basta mudar para ficar assim:
> 
> ```
> import json
> from pynubank import Nubank, MockHttpClient
> 
> # nu = Nubank(MockHttpClient())
> nu = Nubank() # Sem o MockHttpCLient :)
> nu.authenticate_with_cert('cpf', 'pass', 'cert')
> 
> # Obter apenas a primeira página de transações
> feed = nu.get_account_feed()
> 
> # Salvar a primeira página em um arquivo JSON
> with open('feed_1.json', 'w') as f:
>     json.dump(feed, f)
> ```

Desculpe pelo erro tão bobo, acabei não reparando o MockHttpClient, pois estava no automático, muito obrigado pela correção! 

> O controle não é feito de maneira assincrona, o YNAB representa exatamente o estado atual de todas as contas num lugar só, mas além disso esse dinheiro está categorizado. Por exemplo, tenho R$ 500 reais na conta e estou com vontade de pedir um ifood. Se eu olho o saldo da conta, eu tenho dinheiro suficiente para isso, mas eu tenho que lembrar de todas as contas que eu ainda preciso pagar. Então se eu olhar pro meu orçamento no YNAB eu consigo ter uma visão clara do que eu posso / não posso fazer.

Usei o termo assíncrona para me referir ao fato de que, o que organizamos no YNAB não se reflete no Nubank e se mesmo assim era tranquilo para o seu uso, no sentido de você saber que precisaria sempre checar primeiro no YNAB e apenas depois disso, fazer a transação de fato, mas pela continuação da sua resposta, entendi que sim, está funcionando e possibilitando um melhor controle financeiro para você.

O Nubank até teve a função organizador de gastos (implementada depois de adquirir a Olivia), mas resolveu tirar do aplicativo e deixar todo mundo sem alternativa.

Você ainda está usando o seu próprio projeto (br-to-ynab) para fazer a sincronização? Pergunto isso, pois ele não teve nenhum movimento nos últimos tempos mesmo você deixando claro que ainda usa o YNAB e, também, pois eu gostaria de testá-lo, mas não estou conseguindo configurar corretamente, mesmo já tendo instalado os pacotes etc.

Obrigado pela atenção e disposição!

Desculpe-me pelo incômodo!

**Comment by pecampelo** [comment]
Já utilizei também buscando 4 vezes ao dia por meses para múltiplas contas, funciona fazendo suas adequações de estrutura de dados e possíveis casos em que há cartões de crédito.

**Reply by davidborghi** [reply]
Obrigado pela resposta!

Você se importaria de compartilhar qual é/era seu fluxo?

O objetivo era controle da vida financeira ou algum outro projeto específico?

Pergunto isso pois estou tentando chegar na melhor solução possível para meus controles de gastos. 

Obrigado pela atenção e disposição!

Desculpe-me pelo incômodo!
